backend/src/main/java/com/bakdata/conquery/models/messages/namespaces/specific/CollectColumnValuesJob.java

```diff
@@ -0,0 +1,103 @@
+import com.bakdata.conquery.models.events.stores.root.StringStore;
+import com.bakdata.conquery.models.messages.namespaces.ActionReactionMessage;
+import com.bakdata.conquery.models.messages.namespaces.NamespacedMessage;
+import com.bakdata.conquery.models.messages.namespaces.WorkerMessage;
+import com.bakdata.conquery.models.query.FilterSearch;
+import com.bakdata.conquery.models.worker.Namespace;
+import com.bakdata.conquery.models.worker.Worker;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * This Job collects the distinct values in the given columns and returns a {@link RegisterColumnValues} message for each column to the namespace on the manager.
+ */
+@Slf4j
+@RequiredArgsConstructor(onConstructor_ = @JsonCreator)
+@CPSType(id = "COLLECT_COLUMN_VALUES", base = NamespacedMessage.class)
+public class CollectColumnValuesJob extends WorkerMessage implements ActionReactionMessage {
+
+	@Getter
+	@NsIdRefCollection
+	private final Set<Column> columns;
+
+	/**
+	 * This exists only on the manager for the afterAllReaction.
+	 */
+	@JsonIgnore
+	private final Namespace namespace;
+
+
+	@Override
+	public void react(Worker context) throws Exception {
+		final Map<Table, List<Bucket>> table2Buckets = context.getStorage().getAllBuckets().stream()
+															  .collect(Collectors.groupingBy(Bucket::getTable));
+
+
+		final ListeningExecutorService jobsExecutorService = MoreExecutors.listeningDecorator(context.getJobsExecutorService());
+
+
+		final List<? extends ListenableFuture<?>> futures =
+				columns.stream()
+					   .map(column ->
+									jobsExecutorService.submit(() -> {
+										final List<Bucket> buckets = table2Buckets.get(column.getTable());
+
+										if (buckets == null) {
+											log.debug("Skipping column '{}' because there are no buckets imported", column);
+											return;
+										}
+
+										final Set<String> values = buckets.stream()
+																		  .flatMap(bucket -> ((StringStore) bucket.getStore(column)).streamValues())
+																		  .collect(Collectors.toSet());
+										context.send(new RegisterColumnValues(getMessageId(), context.getInfo().getId(), column, values));
+									})
+					   )
+					   .collect(Collectors.toList());
+
+
+		final ListenableFuture<List<Object>> all = Futures.allAsList(futures);
+
+		while (true) {
+			try {
+				all.get(30, TimeUnit.SECONDS);
+				break;
+			}
+			catch (ExecutionException e) {
+				throw new RuntimeException(e);
+			}
+			catch (TimeoutException e) {
+				log.debug("Still waiting for jobs.");
+			}
+		}
```

Comment:
das sieht so falsch aus ._. machen wir das im SeralizingStore auch so?

Reply:
Ich schaue es mir nochmal an, habs bisher einfach rüberkopiert

Reply:
Ja die Statistiken wären wichtig zu loggen. Mir geht es einfach darum, dass es halt Exception-Oriented Programming ist, was mich sehr irritiert.
----

Gerade nachgeschaut, da machen wir das auch so, haben aber noch in der while Bedingung `while(jobs.isDone())` das ist ändert nichts am eigentlichen flow, löst aber die Ästhetischen Bedenken :D Dann kann das break nämlich weg.

Reply:
Ich zähle jetzt mit wer fertig ist